WcaOnRails/spec/requests/registrations_spec.rb

```diff
@@ -430,6 +430,212 @@
+            payment_method_id: pm.id,
+            amount: registration.outstanding_entry_fees.cents,
+          }
+          stripe_charge_id = registration.registration_payments.first.stripe_charge_id
+          stripe_charge = StripeCharge.find_by(stripe_charge_id: stripe_charge_id)
+          expect(stripe_charge&.status).to eq "success"
+          metadata = JSON.parse(stripe_charge.metadata)["metadata"]
+          expect(metadata["wca_id"]).to eq registration.user.wca_id
+        end
+      end
+
+      context "with a valid 3D-secure credit card" do
+        let!(:card) { FactoryBot.create(:credit_card, :sca_card) }
+        let!(:pm) {
+          # This emulates the client side action from the user: when submitting a payment
+          # Stripe creates a payment method and POST its id to us.
+          Stripe::PaymentMethod.create(
+            { type: "card", card: card },
+            stripe_account: competition.connected_stripe_account_id,
+          )
+        }
+
+        it "asks for further action before recording payment" do
+          original_outstanding_fees = registration.outstanding_entry_fees
+          expect(original_outstanding_fees).to be > 0
+          post registration_payment_intent_path(registration.id), params: {
+            payment_method_id: pm.id,
+            amount: original_outstanding_fees.cents,
+          }
+          # The #process_payment_intent endpoint doesn't redirect, it's
+          # the 'register' page which does.
+          expect(registration.reload.outstanding_entry_fees).to eq original_outstanding_fees
+          expect(registration.paid_entry_fees).to eq 0
+          response_json = JSON.parse(response.body)
+          expect(response_json["requires_action"]).to be true
+          # That's as far as we can go, testing the authentication success/failure
+          # must be done by clicking on a modal.
+        end
+
+        it "inserts an 'intent registered' event in the stripe journal" do
+          expect(StripeCharge.all.length).to eq 0
+          post registration_payment_intent_path(registration.id), params: {
+            payment_method_id: pm.id,
+            amount: registration.outstanding_entry_fees.cents,
+          }
+          stripe_charge = StripeCharge.first
+          expect(stripe_charge&.status).to eq "payment_intent_registered"
+          metadata = JSON.parse(stripe_charge.metadata)
+          expect(metadata["payment_method"]).to eq pm.id
+          expect(metadata["metadata"]["wca_id"]).to eq registration.user.wca_id
+        end
+      end
+
+      # The tests below are to test that our endpoint correctly forwards errors,
+      # not to actually test Stripe's correctness...
+      context "rejected credit cards" do
+        it "rejects payment with declined credit card" do
+          original_outstanding_fees = registration.outstanding_entry_fees
+          expect(original_outstanding_fees).to be > 0
```

Comment:
Hmm, this is actually pretty repetitive and doesn't really have to do with the tested functionality itself. I'd suggest either making it clear somehow that that's the case (i.e. during creation) or running that in a before block.

Reply:
Sure, I've specified an explicit `base_entry_fee_lowest_denomination` when creating the competition.